tutorial/lesson_19_staging_func_or_image_param.cpp

```diff
@@ -0,0 +1,289 @@
+                           x, y, halide_result_f2(x, y), c_result_f2[y][x]);
+                    return -1;
+                }
+            }
+        }
+    }
+
+    {
+        // Func::in() is useful to stage loads from a Func via some intermediate
+        // buffer (perhaps on the stack or in shared GPU memory).
+
+        // Let's say 'f' is really expensive to compute and used in several
+        // places. We don't want to recompute 'f' often, but since it is large,
+        // it does not fit in the cache. To deal with this issue, we can do
+        // the following:
+        Func f("f"), g("g");
+        f(x, y) = x + y;
+        g(x, y) = 2 * f(y, x);
+        // First, compute 'f' at root.
+        f.compute_root();
+        // Then, we use Func::in() to stage the loads from 'f' in tiles
+        // as necessary:
+        Var xi("xi"), yi("yi");
+        g.tile(x, y, xi, yi, 8, 8);
+        f.in(g).compute_at(g, x);
+
+        Buffer<int> halide_result = g.realize(40, 40);
+
+        // The equivalent C is:
+        int f_buffer[40][40];
+        for (int y = 0; y < 40; y++) {
+            for (int x = 0; x < 40; x++) {
+                f_buffer[y][x] = x + y;
+            }
+        }
+
+        int c_result[40][40];
+        for (int y = 0; y < 40; y++) {
+            for (int x = 0; x < 40; x++) {
+                int f_wrapper[8][8];
+                for (int yi = 0; yi < 8; yi++) {
+                    for (int xi = 0; xi < 8; xi++) {
+                        f_wrapper[y][x] = f_buffer[y][x];
+                    }
+                }
+                for (int yi = 0; yi < 8; yi++) {
+                    for (int xi = 0; xi < 8; xi++) {
+                        c_result[y][x] = 2 * f_wrapper[y][x];
+                    }
+                }
+            }
+        }
+
+        // Check the results match:
+        for (int y = 0; y < 40; y++) {
+            for (int x = 0; x < 40; x++) {
+                if (halide_result(x, y) != c_result[y][x]) {
+                    printf("halide_result(%d, %d) = %d instead of %d\n",
+                           x, y, halide_result(x, y), c_result[y][x]);
+                    return -1;
+                }
+            }
+        }
+    }
+
+    {
+        // Func::in() can also be used to group multiple stages of a Func
+        // in the same loopness. Consider the following code:
+        Func f("f"), g("g");
+        f(x, y) = x + y;
+        f(x, y) += x - y;
+        g(x, y) = x * y * f(x, y);
+
+        // When we schedule 'f' to be computed at root (by calling f.compute_root()),
+        // all its stages are computed at separate loopness:
+        // for y:
+        //   for x:
+        //     f(x, y) = x + y
+        // for y:
+        //   for x:
+        //     f(x, y) += x - y
+        // for y:
+        //   for x:
+        //     g(x, y) = x * y * f(x, y)
+
+        // We can use Func::in() to group those stages to be computed at the
+        // same loopness like so:
```

Comment:
I think it's worth adding a sentence here explaining *why* this works (by saying how, exactly, it does that). It's not immediately obvious that this is semantically equivalent to applying the inline reduction trick (a trick which is itself not previously explained in the tutorials).